packages/test/test-utils/src/loaderContainerTracker.ts

```diff
@@ -311,48 +312,78 @@ export class LoaderContainerTracker implements IOpProcessingController {
 	 *
 	 * @param containersToApply - the set of containers to check
 	 */
-	private isSequenceNumberSynchronized(containersToApply: IContainer[]) {
+	private needSequenceNumberSynchronize(containersToApply: IContainer[]) {
+		// If there is a pending proposal, wait for it to be accepted
+		const minSeqNum = containersToApply[0].deltaManager.minimumSequenceNumber;
+		if (minSeqNum < this.lastProposalSeqNum) {
+			return {
+				reason: "Proposal",
+				message: `waiting for MSN to advance to proposal at sequence number ${this.lastProposalSeqNum}`,
+			};
+		}
+
 		// clientSequenceNumber check detects ops in flight, both on the wire and in the outbound queue
 		// We need both client sequence number and isDirty check because:
 		// - Currently isDirty flag ignores ops for task scheduler, so we need the client sequence number check
 		// - But isDirty flags include ops during forceReadonly and disconnected, because we don't submit
 		//   the ops in the first place, clientSequenceNumber is not assigned
 
-		const isClientSequenceNumberSynchronized = containersToApply.every((container) => {
+		const containerWithInflightOps = containersToApply.filter((container) => {
 			if (container.deltaManager.readOnlyInfo.readonly === true) {
 				// Ignore readonly container. the clientSeqNum and clientSeqNumObserved might be out of sync
 				// because we transition to readonly when outbound is not empty or the in transit op got lost
-				return true;
+				return false;
 			}
 			// Note that in read only mode, the op won't be submitted
 			let deltaManager = container.deltaManager as any;
 			// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
 			const { trailingNoOps } = this.containers.get(container)!;
-			// Back-compat: clientSequenceNumber & clientSequenceNumberObserved moved to ConnectionManager in 0.53
+			// Back-compat: lastSubmittedClientId/clientSequenceNumber/clientSequenceNumberObserved moved to ConnectionManager in 0.53
 			if (!("clientSequenceNumber" in deltaManager)) {
 				deltaManager = deltaManager.connectionManager;
 			}
 			assert("clientSequenceNumber" in deltaManager, "no clientSequenceNumber");
 			assert("clientSequenceNumberObserved" in deltaManager, "no clientSequenceNumber");
+			// If last submittedClientId isn't the current clientId, then we haven't send any ops
 			return (
-				deltaManager.clientSequenceNumber ===
-				(deltaManager.clientSequenceNumberObserved as number) + trailingNoOps
+				deltaManager.lastSubmittedClientId === container.clientId &&
+				deltaManager.clientSequenceNumber !==
+					(deltaManager.clientSequenceNumberObserved as number) + trailingNoOps
 			);
 		});
 
-		if (!isClientSequenceNumberSynchronized) {
-			return false;
+		if (containerWithInflightOps.length !== 0) {
```

Comment:
Is there a behavior change in this function? or is just refactor?

Reply:
Yes. This looked wrong to me before, because it would only wait for inbound ops once, and then go into a busy wait.  I think the intention was to limit the console noise only, so I change it so that we will wait for inbound ops if we need to.  

I also change it so that a more precise reason can be shown by getting that information from `needSequenceNumberSynchronize`